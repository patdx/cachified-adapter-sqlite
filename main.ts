import { type Cache, totalTtl } from '@epic-web/cachified'
import type { Client } from '@libsql/client'

type KvCacheRow = {
  key: string
  value: string
  metadata: string
  expires_at: number | null
  created_at: number | null
}

const initSql = [
  `create table if not exists kv_cache (
    key text primary key,
    value text not null,
    metadata text not null,
    expires_at integer,
    created_at integer
  ) strict;`,
  `create index if not exists kv_cache_expires_at on kv_cache(expires_at);`,
  `create index if not exists kv_cache_created_at on kv_cache(created_at);`,
]

export async function initDb(db: Client) {
  await db.batch(initSql.map((sql) => ({ sql, args: [] })))
}

export function createCache(db: Client): Cache {
  return {
    name: 'cachified-adapter-sqlite',
    set(key, value) {
      const ttl = totalTtl(value?.metadata)
      const created_at = value.metadata.createdTime
      const expires_at =
        ttl > 0 && ttl < Infinity && typeof created_at === 'number'
          ? (created_at + ttl)
          : null

      return db.execute({
        sql:
          `insert or replace into kv_cache (key, value, metadata, expires_at, created_at)
                                    values (?, ?, ?, ?, ?)`,
        args: [
          key,
          JSON.stringify(value.value),
          JSON.stringify(value?.metadata),
          expires_at,
          created_at,
        ],
      })
    },
    async get(key) {
      const rows = await db.execute({
        sql: 'SELECT value, metadata FROM kv_cache WHERE key = ?',
        args: [key],
      })

      const row = rows.rows?.[0] as unknown as KvCacheRow | undefined

      if (!row) return null

      return {
        value: JSON.parse(row.value),
        metadata: JSON.parse(row.metadata),
      }
    },
    delete(key) {
      return db.execute({
        sql: 'DELETE FROM kv_cache WHERE key = ?',
        args: [key],
      })
    },
  }
}

export async function clearExpired(db: Client) {
  await db.execute({
    sql: 'DELETE FROM kv_cache WHERE expires_at < ?',
    args: [Date.now()],
  })
}
