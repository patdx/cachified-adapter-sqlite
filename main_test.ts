import { createClient } from '@libsql/client/node'
import type { Client } from '@libsql/client/web'
import { assertSnapshot } from '@std/testing/snapshot'
import { FakeTime } from '@std/testing/time'
import { createCache, initDb } from './main.ts'
import { cachified, verboseReporter } from '@epic-web/cachified'

const db = createClient({
	url: ':memory:',
})

Deno.test('init db', async (ctx) => {
	using time = new FakeTime(1000000)
	await initDb(db)

	await assertSnapshot(ctx, await checkSchema(db))

	const cache = createCache(db)

	await cachified({
		cache,
		key: 'hello',
		getFreshValue: async () => 'hello',
		ttl: 1000,
		swr: 1000,
	}, verboseReporter())

	const metadata = await db.execute({
		sql: `select value, metadata from kv_cache where key = ?`,
		args: ['hello'],
	})

	await assertSnapshot(ctx, metadata)
})

async function checkSchema(db: Client) {
	const result = await db.execute({
		sql: `select * from sqlite_schema`,
		args: [],
	})

	return result
}
