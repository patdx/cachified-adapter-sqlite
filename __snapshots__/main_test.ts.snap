export const snapshot = {};

snapshot[`init db 1`] = `
ResultSetImpl {
  columnTypes: [
    "TEXT",
    "TEXT",
    "TEXT",
    "INT",
    "TEXT",
  ],
  columns: [
    "type",
    "name",
    "tbl_name",
    "rootpage",
    "sql",
  ],
  lastInsertRowid: undefined,
  rows: [
    {
      name: "kv_cache",
      rootpage: 2,
      sql: "CREATE TABLE kv_cache (
    key text primary key,
    value text not null,
    metadata text not null,
    expires_at integer,
    created_at integer
  ) strict",
      tbl_name: "kv_cache",
      type: "table",
    },
    {
      name: "sqlite_autoindex_kv_cache_1",
      rootpage: 3,
      sql: null,
      tbl_name: "kv_cache",
      type: "index",
    },
    {
      name: "kv_cache_expires_at",
      rootpage: 4,
      sql: "CREATE INDEX kv_cache_expires_at on kv_cache(expires_at)",
      tbl_name: "kv_cache",
      type: "index",
    },
    {
      name: "kv_cache_created_at",
      rootpage: 5,
      sql: "CREATE INDEX kv_cache_created_at on kv_cache(created_at)",
      tbl_name: "kv_cache",
      type: "index",
    },
  ],
  rowsAffected: 0,
}
`;

snapshot[`init db 2`] = `
ResultSetImpl {
  columnTypes: [
    "TEXT",
    "TEXT",
  ],
  columns: [
    "value",
    "metadata",
  ],
  lastInsertRowid: undefined,
  rows: [
    {
      metadata: '{"ttl":1000,"swr":1000,"createdTime":1000000}',
      value: '"hello"',
    },
  ],
  rowsAffected: 0,
}
`;
